File: merge sort/time_merge_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <ctime>
#include <cstdlib>
#include <fstream>


using namespace std;

void merge(vector<int>& arr, int left, int mid, int right) {
    int left_size = mid - left + 1;
    int right_size = right - mid;

    vector<int> left_half(left_size), right_half(right_size);

    for (int i = 0; i < left_size; ++i)
        left_half[i] = arr[left + i];

    for (int j = 0; j < right_size; ++j)
        right_half[j] = arr[mid + 1 + j];

    int i = 0, j = 0, k = left;

    while (i < left_size && j < right_size) {
        if (left_half[i] <= right_half[j]) {
            arr[k] = left_half[i];
            ++i;
        }
        else {
            arr[k] = right_half[j];
            ++j;
        }
        ++k;
    }

    while (i < left_size) {
        arr[k] = left_half[i];
        ++i;
        ++k;
    }

    while (j < right_size) {
        arr[k] = right_half[j];
        ++j;
        ++k;
    }
}

void merge_sort(vector<int>& arr, int left, int right) {
    if (left >= right)
        return;

    int mid = left + (right - left) / 2;
    merge_sort(arr, left, mid);
    merge_sort(arr, mid + 1, right);
    merge(arr, left, mid, right);
}

double measure_time_sorting(int list_size) {
    // Generar lista de números desordenados aleatorios
    vector<int> arr(list_size);
    srand(time(0));

    for (int i = 0; i < list_size; ++i)
        arr[i] = rand() % 100000 + 1;

    // Medir tiempo de ejecución
    clock_t start_time = clock();
    merge_sort(arr, 0, list_size - 1);
    clock_t end_time = clock();

    return static_cast<double>(end_time - start_time) / CLOCKS_PER_SEC;
}

void export_to_csv_cpp(const vector<pair<int, double>>& results) {
    ofstream csvfile("results_cpp.csv");

    csvfile << "Tamaño de la lista, Tiempo de ejecución (segundos)" << endl;

    for (const auto& result : results) {
        csvfile << result.first << ", " << result.second << endl;
    }

    csvfile.close();
}

int main() {
    vector<int> list_sizes = {100, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000, 20000, 30000, 40000, 50000};

    cout << "Tiempos de ejecución para diferentes tamaños de entrada:" << endl;
    cout << "| Tamaño de la lista | Tiempo de ejecución (segundos) |" << endl;
    cout << "|--------------------|-------------------------------|" << endl;
    vector<pair<int, double>> results;
    for (int size : list_sizes) {
        double execution_time = measure_time_sorting(size);
        cout << "| " << size << "\t\t  | " << execution_time << "\t\t   |" << endl;
        results.push_back(make_pair(size, execution_time));
    }
    export_to_csv_cpp(results);

    return 0;
}

File: merge sort/time_merge_sort.py
import random
import time
import csv

def export_to_csv_python(results):
    with open('results_python.csv', 'w', newline='') as csvfile:
        fieldnames = ['Tamaño de la lista', 'Tiempo de ejecución (segundos)']
        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
        writer.writeheader()
        
        for size, time in results:
            writer.writerow({'Tamaño de la lista': size, 'Tiempo de ejecución (segundos)': time})

def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left_half = merge_sort(arr[:mid])
    right_half = merge_sort(arr[mid:])

    return merge(left_half, right_half)

def merge(left, right):
    merged = []
    left_idx, right_idx = 0, 0

    while left_idx < len(left) and right_idx < len(right):
        if left[left_idx] < right[right_idx]:
            merged.append(left[left_idx])
            left_idx += 1
        else:
            merged.append(right[right_idx])
            right_idx += 1

    merged.extend(left[left_idx:])
    merged.extend(right[right_idx:])
    return merged

def measure_time_sorting(list_size):
    # Generar lista de números desordenados aleatorios
    arr = [random.randint(1, 100000) for _ in range(list_size)]

    # Medir tiempo de ejecución
    start_time = time.time()
    sorted_arr = merge_sort(arr)
    end_time = time.time()

    return end_time - start_time

if __name__ == "__main__":
    list_sizes = [100, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000, 20000, 30000, 40000, 50000]

    print("Tiempos de ejecución para diferentes tamaños de entrada:")
    print("| Tamaño de la lista | Tiempo de ejecución (segundos) |")
    print("|--------------------|-------------------------------|")
    results = []
    for size in list_sizes:
        execution_time = measure_time_sorting(size)
        results.append((size, execution_time))
        print(f"| {size:19} | {execution_time:29} |")
    print("Resultados almacenados en la lista 'results':")
    print(results)
    #export_to_csv_python(results)
    export_to_csv_python(results)
